Feat/mini challenge 5

PedroEdu6786 opened the pull request:
## PR Summarize
- Add authentication with auth0
- Add auth0 login view
- New add/remove favorites button feature
- Add favorite video details view
- Add favorites view
- Add protected route component
- Add tests for each new component created

## Test Coverage
![image](https://user-images.githubusercontent.com/50753661/130706058-c77e373d-ce69-4976-865f-f9c6b6910d3d.png)

PedroEdu6786 committed:
:sparkles: feat: add new HomePage screen with dynamic video list and styles

Add styled-components and react-icons library, new header, layout, homepage and videopreview components, add styled-components theme to project based on chakra ui library

PedroEdu6786 committed:
:white_check_mark: feat: add test for header, previewlist and homepage components

Add test coverage report in coverage.txt, implemented tests for rendering

PedroEdu6786 committed:
:lipstick: fix: background color on VideoPreview component

fixed background color change on some VideoPreview cards

PedroEdu6786 committed:
:art: feat: add code review refactoring PR 1

Extracted map function into outside renderVideoItem function

PedroEdu6786 committed:
:sparkles: feat: add useYoutubeApi hook, dynamic videoDetails page

Add dynamic links to videodetails page, add new hook for youtube's api data fetching, add new axios dependency, change header drawer link to home using react-router, add react-router navigation, refactor styles

PedroEdu6786 committed:
:sparkles: feat: dynamic videos from youtube api, useContext state

Add useContext state management, useGlobalState hook, fetch videos from youtube api

PedroEdu6786 committed:
:sparkles: feat: optimize home page using globalstate

update youtubeapi hook to update globalstate, homepage now uses globalstate videos instead of useYoutubeApi data

PedroEdu6786 committed:
:white_check_mark: feat: refactor tests

Pass all test for all test files, updated coverage.txt for more details

PedroEdu6786 committed:
:white_check_mark: feat: Add global context to home test

refactor test for homepage and remove unused commented code on videodetails

PedroEdu6786 committed:
:white_check_mark: feat: add youtubeApi hook tests, fix homepage render test warning

Add new useYoutubeApi hook tests, add new dependency for hook testing, fix warnings on homepage tests run

PedroEdu6786 committed:
:recycle: refactor: useGlobalState, GlobalContextProvider

refactor globalStateProvider to videosProvider, created videosProvider folder structure, refactor useGlobalState for reusability for every provider

PedroEdu6786 committed:
:sparkles: feat: add theme toggle, add dark theme

Add ThemeContentProvider for theme global context management, add toggle theme to header

PedroEdu6786 committed:
:white_check_mark: feat: add tests for useYoutubeApi and helpers

update useYoutubeApi hook test with axios mock functions, add test for failing case on useYoutubeApi. Add tests for helper functions (youtubeClient and buildQueryParams)

PedroEdu6786 committed:
:white_check_mark: feat: Add useGlobalState tests

add new useGlobalState tests for: default state, update state and default dispatch case

PedroEdu6786 committed:
:white_check_mark: feat: update homepage test

fix homepage test warnings, add test for data fetching from api

PedroEdu6786 committed:
:white_check_mark: feat: add test for reducers

add test cases for videosReducer and themeReducer

PedroEdu6786 committed:
:white_check_mark: feat: add Header and PreviewRelatedList tests

Add test for theme toggle on header component. Add test for render related video data

PedroEdu6786 committed:
:recycle: feat: add code review changes from 2nd delivery

move header styles to styles file, add unit testing for overlay on header, new test for renderVideoItem

PedroEdu6786 committed:
:recycle: feat: VideoProvider and ThemeContextProvider

Change file name extension to keep consistency, add actions file for each provider, add changes for files affected

PedroEdu6786 committed:
:white_check_mark: feat: add test for homepage error heading

Add render test for error heading when api request fails

PedroEdu6786 committed:
:recycle: feat: youtubeHelpers file

Move youtubeClient to youtubeHelpers file, change files affected by refactor

PedroEdu6786 committed:
:heavy_plus_sign: feat: add auth0 dependency

PedroEdu6786 committed:
:sparkles: feat: add authentication with auth0

Add authentication button to header, setup auth0 on index.jsx, toggle logout and login on header when authenticated, add new GhostButton styled component

PedroEdu6786 committed:
:sparkles: feat: add FavoriteVideos page

Add new FavoriteVideos page, add to favorites button VideoDetails page with auth, add new styles for page created

PedroEdu6786 committed:
:sparkles: feat: add Private routes

add private routes component, add routing to favorites page, redirect to home if not authenticated

PedroEdu6786 committed:
:recycle: refactor: move data to VideoDetailsContent

move videodetails main video to VideoDetailsContent for reusability on FavoriteVideoDetais

PedroEdu6786 committed:
:sparkles: feat set_favorite_videos reducer, VideoProvider

add reducer and action to set localStorage videos

PedroEdu6786 committed:
:sparkles: feat: add FavoriteVideoDetails page

change to PreviewRelatedList for dynamic paths, add FavoriteVideoDetails page, add styles for atoms, add dispatcher for setFavoriteVideosAction on VideoDetails and FavoriteVideoDetails, change to home for dynamic paths, add FavoriteVideoDetails path to App component

PedroEdu6786 committed:
:sparkles: feat: add remove video from favorites

add remove from favorites implementation, add action and reducer for video remove from favorites for videoprovider

PedroEdu6786 committed:
:white_check_mark: feat: add VideoDetailsContent tests

test: add video to favorites list, renders favorite button only when authenticated, removes video from favorites

PedroEdu6786 committed:
:white_check_mark: feat: helpers test cases, refactor helper files

add tests for video and localStorage helpers, change files affected by helpers file refactor

PedroEdu6786 committed:
:white_check_mark: feat: add test for FavoriteVideoDetails

add tests for render and render PreviewVideoList

PedroEdu6786 committed:
:white_check_mark: feat: add tests for FavoriteVideos page

add tests for rendering and for preview videos list

PedroEdu6786 committed:
:recycle: refactor: useEffect getFromLocalStorage

refactor getFromLocalStorage functionality on useEffect for localStorage helper

PedroEdu6786 committed:
:lipstick: fix: update styles from home and favoriteVideos

update alignment from favoriteVideos and home page styles

PedroEdu6786 committed:
:white_check_mark: feat: add test for ThemeContentProvider

add test for reducer default case

PedroEdu6786 committed:
:white_check_mark: feat: add test for VideoProvider

add test for action SET_FAVORITE_VIDEOS and default case

PedroEdu6786 committed:
:sparkles: feat: add NotFound page

add redirect to NotFound page when using unknown path, add NotFound page styles and tests

PedroEdu6786 committed:
:white_check_mark: feat: add App navigation test

add navigation test from home to VideoDetails

PedroEdu6786 committed:
:lipstick: update styles file

PedroEdu6786 committed:
:white_check_mark: feat: add test for private routes

JorgeMHC-dev commented on src/components/Search/Search.component.test.jsx:
```diff
@@ -0,0 +1,42 @@
+    youtubeClient.mockImplementationOnce(() => Promise.resolve({ data: videos }));
+    const { input } = build();
+
+    await waitFor(() => {
```
Que onda @PedroEdu6786 el waitFor solo necesita tener dentro el expect, el fire event lo puedes dejar fuera, eso es solo para esperar a que se cambie el state de tu componente 😄

PedroEdu6786 replied:
Oh ya veo, realizaré la modificación

PedroEdu6786 replied:
Igual te invito a observar el proyecto funcional en el siguiente enlace: https://react-certification-challenge-5.netlify.app/

JorgeMHC-dev commented on src/utils/helpers/helpers.test.js:
```diff
@@ -0,0 +1,73 @@
+      };
+
+      youtubeClient.mockImplementationOnce(() => Promise.resolve(response));
+
```
Creo que te hizo falta crear el mock de la funcion de youtube para usar el mockImplementationOnce, la prueba falla, pero se ve bien lo demas 👌

PedroEdu6786 replied:
@JorgeMHC-dev Dejé un comentario acerca de esa situación dentro del readme en codesandbox, parece ser problema con codesandbox lo que provoca que surja el error, te invito a descargar el proyecto localmente para correr las pruebas

JorgeMHC-dev reviewed (commented):
Que onda! @PedroEdu6786 Esta vez me toco a mi revisar el PR jajajaja espero eso no te confunda, revisando el codigo, la neta se ve bastante bien! debo admitir que tienes muy buen manejo de react, testing y los demás temas, te deje unos pequeños comments que realmente no es mucho problema en tus tests y no tambien, no pude ver la app funcionando en si, hizo falta el api key de youtube y me daba error, como recomendacion, podrias usar un fallback por si falla ese request usar el archivo de mocked videos solo para mostrar algo y poner algun aviso o algo que son los videos de mocked que fallo el request a youtube por la razon que tiro, pero de ahi en mas todo perfecto! tratare de sacar una api key propia para ver funcionar el proyecto! sigue asi! cualquier duda puedees buscarme en slack soy Jorge Hernandez Camacho

JorgeMHC-dev commented:
@PedroEdu6786  Una disculpa ya vi el comentario que pusiste diciendo que no iba a funcionar jajajajaja ya vi como hacerlo funcionar 😅

PedroEdu6786 committed:
:sparkles: feat: add login modal view

Add modal view using auth0 api loginWithPopup function

PedroEdu6786 committed:
:white_check_mark: feat: move fireEvent outside waitFor

Recommendation from mini challenge 5 pull request

PedroEdu6786 committed:
:bug: fix: add loading for private routes

Loading message when accessing directly to a private route

PedroEdu6786 committed:
:rocket: feat: add redirects to solve netlify 404

Add redirect files for netlify redirects on deploy

PedroEdu6786 committed:
:sparkles: feat: add loader when loading api videos

add loader icon and animation to homepage

PedroEdu6786 committed:
:clown_face: feat: remove channel from videos mock

PedroEdu6786 committed:
:art: update: findVideoById

returns false if array param is null or undefined

PedroEdu6786 committed:
:lipstick: update: change container styles

PedroEdu6786 committed:
:art: feat: add loader for loading state

PedroEdu6786 committed:
:sparkles: feat: add/remove button on video card hover

PedroEdu6786 committed:
:memo: update: documentation README.md file

Add project description, portfolio and deployed projects links, add features and how to clone project description

PedroEdu6786 committed:
:memo: fix: wizeline academy heading style
